runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1423,6 +1535,122 @@ J9::ValuePropagation::doDelayedTransformations()
+      if (isCompare)
+         {
+         continue;
+         }
+
+      TR_ASSERT_FATAL_WITH_NODE(callNode, !comp()->requiresSpineChecks(), "Cannot handle VP yet for jit{Load|Store}FlattenableArrayElement if SpineCHKs are required\n");
+
+      int opIndex = 0;
+
+      TR::Node *valueNode = isLoad ? NULL : callNode->getChild(opIndex++);
+      TR::Node *indexNode = callNode->getChild(opIndex++);
+      TR::Node *arrayRefNode = callNode->getChild(opIndex);
+
+      TR::Node *elementAddressNode = J9::TransformUtil::calculateElementAddress(comp(), arrayRefNode, indexNode, TR::Address);
+
+      if (valueNode != NULL)
+         {
+         valueNode->decReferenceCount();
+         }
+
+      indexNode->decReferenceCount();
+      arrayRefNode->decReferenceCount();
+
+      const int32_t width = comp()->useCompressedPointers() ? TR::Compiler->om.sizeofReferenceField()
+                                                            : TR::Symbol::convertTypeToSize(TR::Address);
+
+      TR::Node *arrayLengthNode = TR::Node::create(callNode, TR::arraylength, 1, arrayRefNode);
+      arrayLengthNode->setArrayStride(width);
+
+      TR::Node *bndChkNode = TR::Node::createWithSymRef(TR::BNDCHK, 2, 2, arrayLengthNode, indexNode,
+                                          comp()->getSymRefTab()->findOrCreateArrayBoundsCheckSymbolRef(comp()->getMethodSymbol()));
+
+      callTree->insertBefore(TR::TreeTop::create(comp(), bndChkNode));
+
+      TR::SymbolReference *elementSymRef = comp()->getSymRefTab()->findOrCreateArrayShadowSymbolRef(TR::Address, arrayRefNode);
+
+      if (isLoad)
+         {
+         const TR::ILOpCodes loadOp = comp()->il.opCodeForIndirectArrayLoad(TR::Address);
+
+         TR::Node *elementLoadNode = TR::Node::recreateWithoutProperties(callNode, loadOp, 1, elementAddressNode, elementSymRef);
+
+         if (comp()->useCompressedPointers())
+            {
+            TR::Node *compressNode = TR::Node::createCompressedRefsAnchor(elementLoadNode);
+            callTree->insertBefore(TR::TreeTop::create(comp(), compressNode));
+            }
+         }
+      else
+         {
+         TR::Node *oldAnchorNode = callTree->getNode();
+         TR::Node *elementStoreNode = TR::Node::recreateWithoutProperties(callNode, TR::awrtbari, 3, elementAddressNode,
+                                                   valueNode, arrayRefNode, elementSymRef);
+
+         if (needsStoreCheck)
+            {
+            TR::ResolvedMethodSymbol *methodSym = comp()->getMethodSymbol();
+            TR::SymbolReference *storeCheckSymRef = comp()->getSymRefTab()->findOrCreateTypeCheckArrayStoreSymbolRef(methodSym);
+            TR::Node *storeCheckNode = TR::Node::createWithRoomForThree(TR::ArrayStoreCHK, elementStoreNode, 0, storeCheckSymRef);
+            callTree->setNode(storeCheckNode);
+            }
+         else
+            {
+            callTree->setNode(elementStoreNode);
+            }
+
+         oldAnchorNode->removeAllChildren();
```

Comment:
I wonder if `prepareToReplaceNode()` should be called instead of `removeAllChildren`

Reply:
I hadn't encountered `prepareToReplaceNode()` before.  I'll look at whether that's more appropriate.

Reply:
> Do we need to decrement the reference count on `oldAnchorNode`?

I don't think we can, because its reference count should already be zero.

> I wonder if `prepareToReplaceNode()` should be called instead of `removeAllChildren`

Looking at it, I don't think `prepareToReplaceNode()` is the right thing to call.  It's used if a node is being recreated with a new opcode, but this code is not doing that with `oldAnchorNode` - it's just discarding it.  I'm trying to figure out whether this should be calling `Optimization::prepareToStopUsingNode()`, instead, or maybe `Optimization::removeNode(TR::Node * node, TR::TreeTop *anchorTree)`. . . .

Reply:
Vijay @vijaysun-omr or Daryl @0xdaryl, I think I need some advice here.  This is removing `oldAnchorNode`, which would be a `TR::treetop` node that was anchoring the call to `jitStoreFlattenableArrayElement`.

As you can see, this code calls `callTree->setNode` to replace that `TR::treetop` node with the newly created `awrtbari` or `ArrayStoreCHK`, and then calls `removeAllChildren()` on the `TR::treetop` node.  But following this [review comment](https://github.com/eclipse-openj9/openj9/pull/12467#discussion_r617916901) about calling [`prepareToReplaceNode()`](https://github.com/eclipse/omr/blob/bb9ac28f71bfa799c6c7effe69a5b3a9e8525f6f/compiler/optimizer/OMROptimization.cpp#L172-L179) on it, I've been rummaging around looking at that and at [`prepareToStopUsingNode()`](https://github.com/eclipse/omr/blob/bb9ac28f71bfa799c6c7effe69a5b3a9e8525f6f/compiler/optimizer/OMROptimization.cpp#L390-L400).

Both of those methods call [`Optimizer::prepareForNodeRemoval`](https://github.com/eclipse/omr/blob/bb9ac28f71bfa799c6c7effe69a5b3a9e8525f6f/compiler/optimizer/OMROptimizer.cpp#L2264-L2306), which is doing various things with use-def info and value number info, and I'm unsure whether that's needed in this situation, or if calling to `removeChildren()` suffices.

I also see various places in value propagation that are calling `invalidateUseDefInfo()` and `invalidateValueNumberInfo()`, but I'm unclear about the conditions under which those should be called, and so whether they should be called in this situation.

Reply:
I believe that what seems to be getting attempted in `prepareForNodeRemoval` is a very rudimentary repair of value numbering and use def info without invalidating completely. However, given the invasive nature of transformations in `doDelayedTransformations` this repair may be pointless anyway since the nature of what goes on in that routine makes it unsuited for repair and so invalidating value numbering and use def info completely is the inevitable end outcome probably. So I personally don't believe it is very important to try and use the api that attempts the repair.

Reply:
Thanks, Vijay @vijaysun-omr.  I've left this using `removeAllChildren()`, rather than attempting to use `prepareForNodeRemoval`.  However, I'm still unclear on the circumstances under which `invalidateUseDefInfo` and `invalidateValueNumberInfo` need to be called, and in particular, whether those methods should be called in this case.

Reply:
If a transformation makes existing use def info or value number info computed before the transformation invalid, usually by creating new loads or stores of autos/temps/parms, then those routines that you mentioned need to be called to signal to the optimizer to rebuild the analysis info from scratch before the next analysis is run.  Optimizations have expectations for finding information about the category of IL nodes/opcodes that they are interested in and if we have some relevant IL nodes participating in use def info (or value number info) but not others, it creates confusion for the optimization to conclude anything and may even result in a crash given the inconsistent/unexpected analysis state.
Note I said all relevant IL nodes for a given analysis; I am not saying that all IL nodes need to participate in use def info for example, they need not and do not. But for IL nodes that are relevant, e.g. loads, stores and anything that could alias them you can imagine how use def info would need to include all relevant nodes. For example, if use def info did not include all loads for a given auto T, then how can dead store elimination be confident to remove a store to auto T even when it did not have any loads/uses as per use def info.

Reply:
Thanks for the explanation, Vijay @vijaysun-omr.  If I'm understanding you correctly, I won't need to call `invalidateUseDefInfo` or `invalidateValueNumberInfo` here.  No new loads or stores of autos, temps or params are introduced and no existing such loads or stores are removed - they're all just reused from the previously existing calls to the `jit{Load|Store}FlattenableArrayElement` helpers or `<objectEqualityComparison>` non-helper.
Similarly, value numbers associated with nodes that are reused still represent the same values after the transformations - the result of the comparison or the array element load/store.

I suppose one case that could be missed by not invalidating the value number info might be the results of two comparisons involving the same operands.  The two calls to `<objectEqualityComparison>` will have different value numbers, which will be preserved by the resultant `acmpeq` operations.  If value number info had been invalidated, subsequent recomputation of the value numbers could show them to represent the same value - but I'm not sure that's an important enough case to justify invalidating the value number info.

I'll mark this as resolved, but please correct me if I'm mistaken in my understanding.